content/posts/building_jnigen.md

```diff
@@ -0,0 +1,262 @@
+type: "post"
+date: 2023-06-29T07:49:21+05:30
+draft: true
+tags: [dart, flutter, jnigen]
+showTableOfContents: true
+---
+>> Caine: Then why bother?
+>>
+>> John: Maybe I am wrong.
+>>
+> John Wick 4, (2023)
+
+It's been a long overdue to write a post about my past year's Google Summer of Code (GSoC) project. `jnigen` is an experimental bindings generator which aims to provide Java interoperability for Dart. It works by generating wrappers which call JNI through Dart's FFI (Foreign Function Interface).
+
+I developed the initial versions of this package under the guidance of Daco Harkes and Liam Appelbe from Dart language team. Hossein Yousefi from Dart team is developing the project further, adding many features such as Generics and Kotlin language support.
+
+## Why?
+
+The current way of accessing platform APIs on Flutter is Method Channels. Channels neatly avoid the intricacies of native language interop, by using an RPC like mechanism instead. However, there are a few drawbacks - all calls to method channels have to be asynchronous, and sharing memory is not possible. Perhaps more important, currently the use of method channels requires writing quite a bit of boilerplate. [^pigeon]
+
+Therefore, it's desirable to have an automatic bindings generator for Java libraries, for both ergonomics reasons. Using FFI instead of method channels enables better performance and eliminates the need for asynchrony as well.
+
+Dart already has an excellent bindings generator for C, called `ffigen`. It was written 2 years ago by Prerak Mann, in 2020 GSoC. It has been later extended to support Objective-C as well. It uses `libclang` to parse the C headers and generate wrapper code for calling them. The goal of `jnigen` was to provide similar facility for Java libraries.
+
+We had the prior art of `ffigen` which simplified several decisions. However there were challenges unique to Java.
+
+- How to parse the Java libraries? In C/C++, libclang is ubiquitous for this purpose.
+- In case of C, the `ffi` library is already provided by the dart. For Java, we had to develop a support library.
+- C dependency graphs are fairly sparse, and you can provide some header file paths to the tool. In comparison Java build systems and dependency configurations tend to be complex systems in their own right.
+
+As a rule of thumb, everything about Java interop (runtime, parsing, build) is less known, or less obvious. As a student, GSoC was my first attempt to write anything with real world scope. Therefore, I learned several practices and principles.
+
+The aim of this post is to give a glance of various architectural and design decisions, as well as the lessons I learned from this project.
+
+__TODO: A High level Architecture Diagram:__
+
+## JNI runtime support
+
+JNI is an old native interface [^jni]. Therefore, lot of quirks have to be considered when our goal is to abstract the JNI to a high level language like Dart.
```

Comment:
I'd remove "old".

In my opinion the quirks are probably necessary. In Dart we have similar restrictions (but isolate instead of thread for callbacks, in java you have to attach thread, in Dart you have to have the current thread entered the right isolate - or use native ports). Also, our `Dart_Handle`s and `Dart_PersisentHandle` in `dart_api.h` are the same as Java's local and global handles. These are just necessary things when being called from C in a managed language.

Reply:
They have a new API in JDK 19. Havent looked deeply into it but its supposed address many shortcomings of JNI.

We wont get it in android anytime soon, though.